fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SizeStatisticsCollectorCursor.java

```diff
@@ -0,0 +1,635 @@
+     *
+     * @param subspace the subspace to collect statistics on key and value sizes
+     * @param context the transaction context under which stats collection occurs
+     * @param scanProperties scan direction, limits, etc. under which underlying record store scan is performed
+     * @param continuation the starting cursor location to start stats collection
+     *
+     * @return a statistics collector of the given subspace
+     */
+    @Nonnull
+    public static SizeStatisticsCollectorCursor ofSubspace(@Nonnull Subspace subspace, @Nonnull FDBRecordContext context,
+                                                           @Nonnull ScanProperties scanProperties, @Nullable byte[] continuation) {
+        return new SizeStatisticsCollectorCursor(new SubspaceProviderBySubspace(subspace), context, scanProperties, continuation);
+    }
+
+    /**
+     * Create an initialized version of size statistics result class.
+     *
+     * @return an initialized version of size statistics result class.
+     */
+    @Nonnull
+    public static SizeStatisticsResults getSizeStatisticsResultsInitialState() {
+        return new SizeStatisticsCollectorCursor.SizeStatisticsResults();
+    }
+
+    /**
+     * Encapsulates the distribution statistics returned by a SizeStatisticsCollectorCursor.
+     */
+    public static class SizeStatisticsResults {
+        private long keyCount;
+        private long keySize;
+        private long maxKeySize;
+        private long valueSize;
+        private long maxValueSize;
+        private long[] sizeBuckets;
+
+        private SizeStatisticsResults() {
+            this.keyCount = 0;
+            this.keySize = 0;
+            this.maxKeySize = 0;
+            this.valueSize = 0;
+            this.maxValueSize = 0;
+            this.sizeBuckets = new long[Integer.SIZE];
+        }
+
+        /**
+         * Update partial kv stats.
+         *
+         * @param kv latest key and value for partial stats update
+         */
+        private void updateStatistics(@Nonnull KeyValue kv) {
+            this.keyCount += 1;
+            this.keySize += kv.getKey().length;
+            this.maxKeySize = Math.max(this.maxKeySize, kv.getKey().length);
+            this.valueSize += kv.getValue().length;
+            this.maxValueSize = Math.max(this.maxValueSize, kv.getValue().length);
+            int totalSize = kv.getKey().length + kv.getValue().length;
+            if (totalSize > 0) {
+                this.sizeBuckets[Integer.SIZE - Integer.numberOfLeadingZeros(totalSize) - 1] += 1;
+            }
+        }
+
+        /**
+         * Sets partial kv stats from a protobuf serialization of partial stats results.
+         *
+         * @param partialResults partial kv stats from a protobuf serialization of partial stats results.
+         */
+        public void setFromProto(RecordCursorProto.SizeStatisticsPartialResults partialResults) {
```

Comment:
I suppose this works. I was kind of expecting this would be a static initializer like on, say, `KeyExpression`.

Reply:
Yeah, this avoids creating a new one. I'll leave it alone. Every touch is a potential screw up.

Reply:
Hm, well, I still kind of think this should be a static `fromProto` rather than a method that overwrites existing data. In particular, it's a little weird that it one can overwrite existing data if one calls it at the wrong time. But I suppose I won't block the PR on it.